dev.zed.Zed.yaml

```diff
@@ -0,0 +1,135 @@
+# dev.zed.Zed.yaml
+app-id: dev.zed.Zed
+runtime: org.freedesktop.Sdk
+runtime-version: '23.08'
+sdk: org.freedesktop.Sdk
+command: zed
+separate-locales: false
+finish-args:
+  # Grant access to Direct Rendering Infrastructure (DRI) for GPU acceleration
+  - --device=dri
+
+  # Allow access to the user's home directory for reading and writing files
+  - --filesystem=home
+
+  # Allow shared memory access for inter-process communication
+  - --share=ipc
+
+  # Enable network access
+  - --share=network
+
+  # Provide access to X11 display server as a fallback
+  - --socket=fallback-x11
+
+  # Enable access to PulseAudio sound server
+  - --socket=pulseaudio
+
+  # Enable access to Wayland display server
+  - --socket=wayland
+
+  # Enable access to Flatpak host commands (Needs linter exception: https://docs.flathub.org/docs/for-app-authors/linter#exceptions)
+  - --talk-name=org.freedesktop.Flatpak
```

Comment:
They relaunch the main CLI binary outside the sandbox to escape it when starting the app. It implicitly puts a dependency on flatpak-spawn as the bus name they must talk to just to start the app. https://github.com/zed-industries/zed/pull/11949#issuecomment-2116541925 https://github.com/zed-industries/zed/blob/fee49fcf657e2514b02b6e9f81003ea48206c231/crates/cli/src/main.rs#L332-L361

Reply:
I think @someone13574 would be a better source for answering this.

From what I understand -- the intent was that Zed wants the experience to feel the same across all packaging methods and formats where a user downloads Zed, opens it, and expects the same commands available to them as they would have in their native non-sandboxed terminal.

While I understand Flathub does have SDKs available for popular languages like Go and Rust -- it will be up to the user to remember and learn to to import and use them properly in the Flatpak environment.

It certainly would be worth a conversation down the line to spawn additional integrated terminals with a Sandbox-only option, but it doesn't seem to be a priority at the moment seeing how the source code itself is ensuring to run host commands.

Reply:
> I think the lack of configurability there is going to limit the usefulness of this sandbox escape. Would be nice to let users configure a specific command, so e.g. Silverblue users can launch it in the toolbox. cc @ ShobuPrime can this be reflected in some issue? It's not a blocker, of course.

There is a environment variable, `ZED_FLATPAK_NO_ESCAPE`, that [disables](https://github.com/zed-industries/zed/blob/fee49fcf657e2514b02b6e9f81003ea48206c231/crates/cli/src/main.rs#L376-L378) the sandbox escape if set to anything, causing the app to be launched from inside the sandbox. Because the Flatpak manifest already includes basic permissions to operate inside the sandbox (which really aren't needed if the escape is used), this option should still have some functionality out of the box (though there will definitely be some things broken).

---

Overall the escape and no escape modes follow these startup sequences:

Escape mode:

0. The cli is run inside a Flatpak (which is done by the desktop entry)
1. Detect that its in a Flatpak by looking for the env variable.
2. Find the flatpak install path on the host system and create a path to the cli.
3. Use execvp to [replace the process](https://docs.rs/exec/latest/exec/fn.execvp.html) with the cli run outside the sandbox. This command also uses `-zed {path the main app flatpak binary}` so it knows to use the flatpak and additionally passes the XDG env vars with the correct flatpak locations.
4. The cli then starts the main app as normal.

Non-escape mode:

0. The cli is run inside a Flatpak
1. Detect that its in a Flatpak by looking for the env variable.
2. See that the `ZED_FLATPAK_NO_ESCAPE` env is set.
3. Set the main binary path to `/app/libexec/zed-editor`.
4. The cli then starts the main binary like normal, but without escaping.

---

If you can think of a better way to make this configurable for users, I'm all ears.